### PR TITLE
Add support for github enterprise

### LINKED ISSUE #329: Support github enterprise in version 2.0 codebase
Describe what problem your feature request solves
The github oauth configuration assumes that you're using github.com and doesn't work with github enterprise.

Describe the solution you'd like
A configuration in .env (or equivalent for prod) that lets you specify a github enterprise url

**Additional context**
#250 was originally created and implemented in #252 but testing was inconclusive and the changes never merged. @jgadsden suggested to implement the feature in the version 2.0 codebase instead to leverage easier development and debuggability.

### PR DESCRIPTION
**Pull request by danielpops**
### Summary ###
Adding support for Github enterprise.  Resurrecting some of what was done in the v1 codebase in #252 and re-implementing in the v2 codebase. Fixes #329 

### Description for the changelog ###
#329 - Adds support for Github enterprise by adding the hostname (optionally port and protocol) to the config

### Testing Done ###
I manually tested with a Github Enterprise Server instance and confirmed that I could log in, view repositories and branches, and create threat models.

### Other info ###
There are a few TODO's that I'd like to address before this ships:
* `td.vue` has a few hard coded references to `www.github.com` in the template files and i'm not sure the best way to plumb through the .env configuration here without just going straight to `process.env` and duplicating logic already present in `td.server`
* Add some unit tests

I also ran into an issue while testing which seems like an issue in the status quo as well. The github API by default only returns the 30 repos and 30 branches. This is a problem in my environment as there are far more repos and branches than that. There probably needs to be some refactoring to support paging in the UI. I would like to avoid coupling that refactor with this change.

### CODE REVIEW

**Comment by jgadsden** [comment]
This is really really good to see - thankyou @danielpops 
The v2-development branch is historical now, we just did not want to remove it, so did you want to rebase it back to main?

**Comment by jgadsden** [comment]
@danielpops did you want to raise an issue on the paging of 30 repos/branches?

**Comment by danielpops** [comment]
> The v2-development branch is historical now, we just did not want to remove it, so did you want to rebase it back to main?

Ha! That's what I get for not paying attention for 6 months 😅 

Ya no worries I'll rebase the changes

**Comment by danielpops** [comment]
> @danielpops did you want to raise an issue on the paging of 30 repos/branches?

Will do


> Also td.vue can ask td.server for the correct URL, I can look at this if you like?

If you can point me to an example of something similar, I'm happy to include it. If it gets more complicated, I'd appreciate your help 😅

> I recall that I struggled to be able to test this last time ... my introductory period with github enterprise ran out ... I can try again 

I'm happy to pair with you on this if you want

**Comment by jgadsden** [comment]
The express server has an api that you can extend at `td.server/src/config/routes.config.js`, probably an unauthenticated API is your best choice - a bit like healthz maybe

File: td.vue/src/views/git/BranchAccess.vue
Comment (jgadsden) [inline td.vue/src/views/git/BranchAccess.vue:6]: possibly access the express server td.server for this value?

File: td.vue/src/views/git/ThreatModelSelect.vue
Comment (jgadsden) [inline td.vue/src/views/git/ThreatModelSelect.vue:8]: the idea you suggested in your pr comments looks a good one to me, access this value from td.server via a new API which are listed in `td.server/src/config/routes.config.js`

**Review by jgadsden — APPROVED** [review]

**Comment by jgadsden** [comment]
Hello @danielpops - do we want to merge this in if it is working? I can not test it myself, and I am happy on you testing it and giving the OK for merge.
What is the effect of the two fix-me?

**Comment by danielpops** [comment]
@jgadsden sorry I hadn't circled back to this yet. For the two fix-me's:

> we can probably mutate this in the vuex store state.repo.selected or similar

I'm not sure how to achieve that OR the previous suggestion of adding an unauthenticated API, but I'm happy to do that with some guidance or pair on it with you or hand this off back to you :)

**Comment by jgadsden** [comment]
Hello @danielpops , I will have a look and see if I can provide an API on the express server to supply the URL

**Comment by jgadsden** [comment]
Hello @danielpops - I have created pull-request #569 'Add server API for provider details' and added you as a reviewer. I think this will do what you need to get the configured hostname from the express server back-end

**Comment by danielpops** [comment]
> I think this will do what you need to get the configured hostname from the express server back-end

@jgadsden I had problems getting this to work, mostly due to my lack of experience with node (I wasn't able to figure out the right way to reference the async API in the vue files). I also had problems iteratively making changes and testing because I had to keep re-building in docker each time which made it difficult for me to try enough things quickly enough to figure it out on my own. I'm happy to ship this as is and hand it off to you to fix the base URL in a follow-up, or if you want to send some hints / diffs I'm happy to incorporate that into my branch. Sorry for the hand holding :(

**Review by jgadsden — APPROVED** [review]
Looks good to merge this, thanks @danielpops 
I am not able to test the GHE access itself, but it does not seem to break anything else, so all good from my side

**Comment by jgadsden** [comment]
Hello @danielpops 
I hope you do not mind, but to merge this I think I need to back out the changes to README.md and env.md
github is saying that the changes are too complicated to merge, and I know we did a lot of work to these two files

Once we merge your changes in then I can reinstate your documentation, if that is OK with you?

**Pull request merged**
